moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -1357,91 +1357,78 @@ bool RobotState::getJacobian(const JointModelGroup* group, const LinkModel* link
     return false;
   }
 
-  if (!group->isLinkUpdated(link->getName()))
+  const std::set<const moveit::core::LinkModel*>& descendant_links = group->getUpdatedLinkModelsSet();
+  if (descendant_links.find(link) == descendant_links.end())
   {
     RCLCPP_ERROR(LOGGER, "Link name '%s' does not exist in the chain '%s' or is not a child for this chain",
                  link->getName().c_str(), group->getName().c_str());
     return false;
   }
 
-  const moveit::core::JointModel* root_joint_model = group->getJointModels()[0];  // group->getJointRoots()[0];
+  const std::vector<const moveit::core::JointModel*>& joint_models = group->getActiveJointModels();
+  if (joint_models.empty())
+  {
+    RCLCPP_ERROR(LOGGER, "The group '%s' doesn't contain any JointModel's. Cannot compute Jacobian.",
+                 group->getName().c_str());
+    return false;
+  }
+  const moveit::core::JointModel* root_joint_model = group->getJointModels().front();
   const moveit::core::LinkModel* root_link_model = root_joint_model->getParentLinkModel();
-  // getGlobalLinkTransform() returns a valid isometry by contract
-  Eigen::Isometry3d reference_transform =
+
+  Eigen::Isometry3d root_pose_world =
       root_link_model ? getGlobalLinkTransform(root_link_model).inverse() : Eigen::Isometry3d::Identity();
   int rows = use_quaternion_representation ? 7 : 6;
   int columns = group->getVariableCount();
-  jacobian = Eigen::MatrixXd::Zero(rows, columns);
-
-  // getGlobalLinkTransform() returns a valid isometry by contract
-  Eigen::Isometry3d link_transform = reference_transform * getGlobalLinkTransform(link);  // valid isometry
-  Eigen::Vector3d point_transform = link_transform * reference_point_position;
-
-  // RCLCPP_DEBUG(LOGGER, "Point from reference origin expressed in world coordinates: %f %f %f",
-  //         point_transform.x(),
-  //         point_transform.y(),
-  //         point_transform.z());
+  jacobian.resize(rows, columns);
 
-  Eigen::Vector3d joint_axis;
-  Eigen::Isometry3d joint_transform;
+  Eigen::Isometry3d root_pose_tip = root_pose_world * getGlobalLinkTransform(link);
+  Eigen::Vector3d tip_point = root_pose_tip * reference_point_position;
```

Comment:
I also think these can be `const`

Reply:
Sorry for all the missing const!
I'm not used to use const on local variables, but I see this is encouraged at PickNik, from this review and others. I'll start using it.